File: C++/416.cpp
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int n = nums.size(), sum = 0;
        for (auto a : nums) sum += a;
        if (sum%2) return false;
        
        sort(nums.rbegin(), nums.rend());
		return helper(nums, 0, sum/2);
    }
private:
	bool helper(vector<int>& nums, int begin, int target) {//注意 回溯如果可以写成这种简单形式就不要写 for+递归形式了
        if (begin == nums.size()) return false;
		if (!target) return true;
		if (nums[begin] > target) return false;
		
		return helper(nums, begin+1, target - nums[begin]) || helper(nums, begin+1, target);
	}
};
__________________________________________________________________________________________________
sample 8844 kb submission
class Solution {
public:
    bool canPartition1(vector<int>& nums) {
        int sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
        }
        if (sum % 2) return false;
        
        sum /= 2;
        
        // 0-1 knapsack with bag size = sum
        int gain[nums.size() + 1][sum + 1];
        
        for (int i = 0; i <= nums.size(); i++) {
            for (int w = 0; w <= sum; w++) {
                if (i == 0 || w == 0) {
                    gain[i][w] = 0;
                } else {
                    gain[i][w] = gain[i-1][w];
                    if (w >= nums[i-1] && gain[i][w] < gain[i-1][w-nums[i-1]] + nums[i-1]) {
                        gain[i][w] = gain[i-1][w-nums[i-1]] + nums[i-1];
                    }
                }
            }
        }
        
        return gain[nums.size()][sum] == sum;
    }
    
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
        }
        if (sum % 2) return false;
        
        sum /= 2;
        
        // 0-1 knapsack with bag size = sum
        int gain[2][sum + 1];
        int r = 0;
        
        for (int i = 0; i <= nums.size(); i++) {
            for (int w = 0; w <= sum; w++) {
                if (i == 0 || w == 0) {
                    gain[r][w] = 0;
                } else {
                    gain[r][w] = gain[1-r][w];
                    if (w >= nums[i-1] && gain[r][w] < gain[1-r][w-nums[i-1]] + nums[i-1]) {
                        gain[r][w] = gain[1-r][w-nums[i-1]] + nums[i-1];
                    }
                }
            }
            r = 1 - r;
        }
        
        return gain[1-r][sum] == sum;
    }
};
__________________________________________________________________________________________________

File: Python3/416.py
__________________________________________________________________________________________________
sample 36 ms submission
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if len(nums) == 1: return False
        s = sum(nums)
        if s % 2 == 1: return False
        t = s//2
        
        for n in nums:
            if n > t: return False
            
        nums.sort(reverse=True)
        
        def search(curr_sum, nums):
            if curr_sum > t: return False
            elif curr_sum == t: return True
            
            for i, n in enumerate(nums):
                new_nums = nums[:i] + nums[i+1:]
                return search(curr_sum + n, new_nums) or search(curr_sum, new_nums)
        
        return search(0, nums)
__________________________________________________________________________________________________
sample 13168 kb submission
class Solution:
    def canFindSum(self, nums, target, ind, n, d):
        if target in d:
            return d[target] 
        if target == 0:
            d[target] = True
        else:
            d[target] = False
            if target > 0:
                for i in range(ind, n):
                    if self.canFindSum(nums, target - nums[i], i+1, n, d):
                        d[target] = True
                        break
        return d[target]
    
    def canPartition(self, nums):
        s = sum(nums)
        if s % 2 != 0: return False
        return self.canFindSum(nums, s/2, 0, len(nums), {})
__________________________________________________________________________________________________
